File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # find middle point
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        # seperate two listNode
        second = slow.next
        slow.next = None
        # reverse second listNode
        pre = None
        while second:
            temp = second.next
            second.next = pre
            pre = second
            second = temp
        # second listNode start pointer is pre.
        first, second = head, pre
        while first and second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        var n = lists.size
        if (n == 0) {
            return null
        }
        if (n == 1) {
            return lists[0]
        }

        var leftList = mergeKLists(lists.take(n / 2).toTypedArray())
        var rightList = mergeKLists(lists.drop(n / 2).toTypedArray())
        
        var dummyNode = ListNode(-1)
        var tail: ListNode = dummyNode
        while (leftList != null && rightList != null) {
            if (leftList.`val` <= rightList.`val`) {
                tail.next = leftList
                leftList = leftList.next
            } else {
                tail.next = rightList
                rightList = rightList.next
            }
            tail = tail.next!!
        }
        while (leftList != null) {
            tail.next = leftList
            leftList = leftList.next
            tail = tail.next!!
        }
        while (rightList != null) {
            tail.next = rightList
            rightList = rightList.next
            tail = tail.next!!
        }
        return dummyNode.next
    }
}
